chromium_src/components/update_client/update_checker.cc

```diff
@@ -0,0 +1,102 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "../../../../components/update_client/update_checker.cc"
+
+namespace update_client {
+
+SequentialUpdateChecker::SequentialUpdateChecker(
+    scoped_refptr<Configurator> config,
+    PersistedData* metadata)
+    : config_(config), metadata_(metadata) {
+      VLOG(3) << "SequentialUpdateChecker";
+    }
+
+SequentialUpdateChecker::~SequentialUpdateChecker() {
+    VLOG(3) << "> ~SequentialUpdateChecker";
+    DCHECK(thread_checker_.CalledOnValidThread());
+    VLOG(3) << "< ~SequentialUpdateChecker";
+}
+
+void SequentialUpdateChecker::CheckForUpdates(
+    const std::string& session_id,
+    const std::vector<std::string>& ids_checked,
+    const IdToComponentPtrMap& components,
+    const base::flat_map<std::string, std::string>& additional_attributes,
+    bool enabled_component_updates,
+    UpdateCheckCallback update_check_callback) {
+  VLOG(3) << "> CheckForUpdates";
+
+  DCHECK(thread_checker_.CalledOnValidThread());
+  DCHECK(!ids_checked.empty());
+
+  for (const auto& app_id : ids_checked) {
+    VLOG(3) << ">  * " << app_id;
+    remaining_ids_.push_back(app_id);
+  }
+
+  session_id_ = session_id;
```

Comment:
Haven't looked through the code much, but is this safe for multiple invocations? For example if an automatic update check is in progress and I got to brave://components page and click a button to check for updates?

Reply:
I think it's safe now, and believe it is also reasonably safe for the future.

Chromium [creates a new `UpdateChecker` for every update request](https://source.chromium.org/chromium/chromium/src/+/master:components/update_client/update_engine.cc;l=157;drc=52c253e300831067aa75506906a4af2064edadb2). So at the moment, when you perform an on-demand update check while an automatic one is in progress, then a new instance is created that doesn't touch the one that is already running. This applies both to Chromium's base implementation and the one in this PR.

From the way `UpdateChecker` is designed and currently used, it seems to me that the intention to have it use throwaway instances is pretty deeply ingrained. Of course, there are no guarantees. But I don't think it is likely to change.

The present implementation tries to introduce as little new code as possible by delegating to `UpdateChecker`. A result of this is that it does spawn a new thread for every request. But these threads are only spawned sequentially, and `SequentialUpdateChecker` hides this fact from its callers. It gathers results internally and fails or succeeds as a whole before returning to the caller. Because it uses several threads and multiple requests, it is more brittle. (If a single request fails, then the entire update check fails.) But I'm not sure there is a better way to work around this. We could perform all requests in a single thread. But then we'd have to duplicate quite a bit more code from `UpdateChecker`.

Reply:
Thanks for the detailed info, @mherrmann! Threads are definitely not an issue. I initially didn't look in the right place and thought the code was creating a new thread for each request concurrently, not sequentially. My only concern was the possibility of reentrancy, but if that can't happen, then it's all good.